src/MessageTrait.php

```diff
@@ -170,8 +178,11 @@ public function withHeader($header, $value)
             );
         }
 
+        $lowerHeader = strtolower($header);
+
         $new = clone $this;
-        $new->headers[$header] = $value;
+        $new->headerNames[$lowerHeader] = $header;
+        $new->headers[$lowerHeader] = $value;
```

Comment:
One argument brought up on the FIG mailing list is that the headers should be returned using the same case as was used to register them. In that case, `$header` would be used to register with the `headers` property. Doing so would also simplify the `getHeaders()` logic (it would just `return $this->headers`), and `getHeader()` would end up looking like:

```php
if (! $this->hasHeader($header)) {
    return null;
}

$header = $this->headerNames[strtolower($header)];
return $this->headers[$header];
```

I'd be willing to merge this patch with that change.